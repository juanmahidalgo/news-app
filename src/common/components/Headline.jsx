import React, { useState } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { makeStyles } from '@material-ui/core/styles';
import {
  Card,
  CardActions,
  CardContent,
  CardMedia,
  IconButton,
  Collapse,
  Typography,
  Tooltip,
} from '@material-ui/core';
import { FacebookSelector, FacebookCounter } from 'react-reactions';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

const useStyles = makeStyles(theme => ({
  card: {
    width: '50%',
    margin: '25px auto',
  },
  media: {
    height: 0,
    paddingTop: '56.25%', // 16:9
  },
  expand: {
    transform: 'rotate(0deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  counter: {
    marginTop: '0.5rem',
  },
}));

const Headline = ({ headline }) => {
  const classes = useStyles();
  const [reactions, setReactions] = useState([
    {
      by: 'Mark Zuckerberg',
      emoji: 'like',
    },
    {
      by: 'Jeff Bezos',
      emoji: 'love',
    },
    {
      by: 'Testing User',
      emoji: 'love',
    }
  ]);
  const [reaction, setReaction] = useState('');
  const [expanded, setExpanded] = useState(false);

  const onSelect = (emoji) => {
    setReactions([...reactions, { by: 'You', emoji }]);
    setReaction(emoji);
  }
  
  const removeReaction = () => {
    setReactions(reactions.filter(reaction => reaction.by !== 'You'));
    setReaction('');
  }

  const handleExpandClick = () => {
    setExpanded(!expanded);
  }

  return (
    <Card className={classes.card}>
      { headline.urlToImage && <CardMedia
        className={classes.media}
        image={headline.urlToImage}
        title="News image"
      />}
      <CardContent>
        <Typography className={classes.title} color="textSecondary" gutterBottom>
          {headline.source.name} - {new Date(headline.publishedAt).toDateString()}
        </Typography>
        <Typography variant="h5" component="h2">
          <a target="_blank" rel="noopener noreferrer" href={headline.url}>
            {headline.title}
          </a>
        </Typography>
        <Typography className={classes.pos} color="textSecondary">
          {headline.description}
        </Typography>
      </CardContent>
      <CardActions>
        <div>
          {!reaction &&
            <FacebookSelector onSelect={onSelect} />
          }
          <div className={classes.counter}>
            <FacebookCounter
              counters={reactions}
              important={['Mark Zuckerberg', 'Jeff Bezos']}
              onClick={removeReaction}
              user="You"
            />
          </div>
        </div>
        <Tooltip title="Read More" aria-label="Read More">
          <IconButton
            className={clsx(classes.expand, {
              [classes.expandOpen]: expanded,
            })}
            onClick={handleExpandClick}
            aria-expanded={expanded}
            aria-label="Show more"
          >
            <ExpandMoreIcon />
          </IconButton>
        </Tooltip>
      </CardActions>
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          {headline.content}
        </CardContent>
      </Collapse>
    </Card>
  );
};

Headline.propTypes = {
  headline: PropTypes.shape({
    url: PropTypes.string,
    urlToImage: PropTypes.string,
    content: PropTypes.string,
    title: PropTypes.string,
    description: PropTypes.string,
    publishedAt: PropTypes.string,
    source: PropTypes.shape({
      name: PropTypes.string,
    }),
  })
}

export default Headline;
