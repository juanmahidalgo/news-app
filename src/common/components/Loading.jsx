import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles(theme => ({
  progress: {
    margin: theme.spacing(2),
  },
  loadingWrapper: {
    alignItems: 'center',
    display: 'flex',
    height: '100%',
    justifyContent: ' center',
  }
}));

const Loading = () => {
  const classes = useStyles();

  return (
    <div className={classes.loadingWrapper}>
      <CircularProgress className={classes.progress} />
    </div>
  );
}


export default Loading;

