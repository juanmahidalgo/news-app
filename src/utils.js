export const success = (res, data) => {
  res.status(200).json({ data, status: 'ok' });
};

export const notFound = (res) => {
  res.status(404);
};

export const returnCode = (res, code) => {
  res.status(code);
};

export const apiURL = 'https://newsapi.org/v2';

export const mockedHeadline = {
  url: 'test',
  urlToImage: 'test',
  content: 'test',
  title: 'test',
  description: 'test',
  publishedAt: 'test',
  source: {
    name: 'test',
  },
};