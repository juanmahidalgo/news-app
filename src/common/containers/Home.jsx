import React, { useEffect, useCallback, Fragment } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { getHeadlines } from '../actions/news';
import Headline from '../components/Headline';
import Loading from '../components/Loading';
import Error from '../components/Error';
import withCategories from '../HOCs/withCategories';

export const Home = () => {
  const dispatch = useDispatch();
  const { headlines, isFetching, error } = useSelector(state => state.news);
  const fetchHeadlines = useCallback(
    () => dispatch(getHeadlines()),
    [dispatch],
  );

  useEffect(() => {
    fetchHeadlines();
  }, []);

  // This is an example on how data should be fetch on the server side rendering
  // we also would need to change this component from a function to a Class extending from React component
  // This works similarly to Next.js's `getInitialProps`
  // static getInitialData() {
  //   return fetchHeadlines();
  //   // return Promise.all([fetchCategories(), fetchHeadlines()]);
  // }

  if (isFetching) {
    return <Loading />
  }

  return (
    <Fragment>
      {
        error
          ? <Error message={'Error while fetching news'} />
          : headlines &&
              headlines.map(
                (headline, index) => (
                  <Headline key={index} headline={headline} />
                )
              )
      }
    </Fragment>
  );
}

export default withCategories(Home);
