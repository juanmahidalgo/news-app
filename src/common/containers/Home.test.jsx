/* global jest */

import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import ReactTestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { createMount } from '@material-ui/core/test-utils';

import { Home } from './Home';
import Loading from '../components/Loading';
import Headline from '../components/Headline';
import { mockedHeadline } from '../../utils';

jest.mock('react-reactions', () => ({
  FacebookSelector: () => null,
  FacebookCounter: () => null,
}))

const mockedGetHeadlines = jest.fn();
const mockedNewsStore = (props) => ({
  headlines: [mockedHeadline, mockedHeadline],
  isFetching: false,
  ...props,
});
const mockStore = configureMockStore([thunk]);

jest.mock('../actions/news', () => ({
  getHeadlines: () => mockedGetHeadlines
}));

describe('<Home />', () => {
  const mount = createMount();

  it('renders correctly', () => {
    const tree = ReactTestRenderer.create(
      <Provider store={mockStore({ news: mockedNewsStore() })}>
        <Home />
      </Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render loading component', () => {
    const wrapper = mount(
      <Provider store={mockStore({ news: mockedNewsStore({ isFetching: true }) })}>
        <Home />
      </Provider>);
    expect(wrapper.find(Loading)).toHaveLength(1);
    expect(wrapper.find(Headline)).toHaveLength(0);
  });

  it('should render 2 headlines', () => {
    const wrapper = mount(
      <Provider store={mockStore({ news: mockedNewsStore() })}>
        <Home />
      </Provider>);
    expect(wrapper.find(Headline)).toHaveLength(2);
  });

  it('should call getHeadlines once finished first render', () => {
    const renderer = ReactTestRenderer.create(
      <Provider store={mockStore({ news: mockedNewsStore() })}>
        <Home />
      </Provider>
    );
    renderer.update();
    expect(mockedGetHeadlines).toHaveBeenCalled();
  });
});
