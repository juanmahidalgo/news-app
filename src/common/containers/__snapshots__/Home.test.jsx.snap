// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
Array [
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
  >
    <div
      className="MuiCardMedia-root makeStyles-media-2"
      style={
        Object {
          "backgroundImage": "url(\\"test\\")",
        }
      }
      title="News image"
    />
    <div
      className="MuiCardContent-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        test
         - 
        Invalid Date
      </p>
      <h2
        className="MuiTypography-root MuiTypography-h5"
      >
        <a
          href="test"
          rel="noopener noreferrer"
          target="_blank"
        >
          test
        </a>
      </h2>
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        test
      </p>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <div>
        <div
          className="makeStyles-counter-5"
        />
      </div>
      <button
        aria-describedby={null}
        aria-expanded={false}
        aria-label="Show more"
        className="MuiButtonBase-root MuiIconButton-root makeStyles-expand-3"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Read More"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
  >
    <div
      className="MuiCardMedia-root makeStyles-media-2"
      style={
        Object {
          "backgroundImage": "url(\\"test\\")",
        }
      }
      title="News image"
    />
    <div
      className="MuiCardContent-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
      >
        test
         - 
        Invalid Date
      </p>
      <h2
        className="MuiTypography-root MuiTypography-h5"
      >
        <a
          href="test"
          rel="noopener noreferrer"
          target="_blank"
        >
          test
        </a>
      </h2>
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        test
      </p>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <div>
        <div
          className="makeStyles-counter-5"
        />
      </div>
      <button
        aria-describedby={null}
        aria-expanded={false}
        aria-label="Show more"
        className="MuiButtonBase-root MuiIconButton-root makeStyles-expand-3"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Read More"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
]
`;
